File: str/reduceStr/reduceString4.java
//Q3: zuma
public String deDup(String input) {
	if(input == null || input.length() <=1)
		return input;
	char[] arr = input.toCharArray();
	int slow = 0;  //INCLUSIVE to the final answer, like index
	int fast = 1;
	for(fast = 1; fast < arr.length; fast++){
		//slow = -1, stack is empty
		//or when no duplicate, we push on to the stack
		//only looks at next one
		if(slow == -1 || arr[fast] != arr[slow])
			arr[++slow] = arr[fast];  //fast not ++!!!
		else{
			//otherwise there is duplicate, pop the element at index cur
			slow--;  //this might cause cur = -1, cur stays at latest char index
			//1  2  2 2
			//s<-sf f f+1  => stack = 1,2...
			while(fast+1 < arr.length && arr[fast] == arr[fast+1])
				fast++;
		}
	}
	return new String(arr, 0, slow+1);  //slow is inclusive
}
print: input, slow, fast for each loop
abbbbcddc
1
1
abbbbcddc
0
4
acbbbcddc
1
5
acdbbcddc
2
6
acdbbcddc
1
7
acdbbcddc
0
8
a%

File: str/reduceStr/reduceString4.py
# solution2: not in place, n time
def removeString4(str):
	res = [None]
	for index, char in enumerate(str):
		if res[-1] == char:
			res.pop()
		else:
			res.append(char)

	return ''.join(res[1:])  # first index is None

# solution2: n^2 time, n space
def removeDuplicates(self, S: str) -> str:
	s=list(S)
	for j in range(len(s)):
		for i in range(1,len(s)):
			if s[i]==s[i-1]:
				s[i],s[i-1]='0','0'
		t=''.join(s)
		t=list(t.replace('0',''))
		if len(s)==len(t):
			break
		else:
			s=t

	return ''.join(t)

def removeString4(input):
	if input is None or not input:
		return input

	arr = list(input)
	slow = 0
	fast = 1

	while fast < len(arr):
		if slow == -1 or arr[slow] != arr[fast]:
			slow += 1
			arr[slow] = arr[fast]
		else:
			while fast + 1 < len(arr) and arr[fast+1] == arr[slow]:
				fast += 1
			slow -= 1

	return ''.join(arr[:slow+1])
